File: Scala/string_algorithms/src/main/scala/suffix_trees/TrieConstruction.scala
package suffix_trees

object TrieConstruction {
  import scala.annotation.tailrec

  private type Node = Int

  final case class EdgeEndpoint(tip: Node, label: Char)

  class Trie(words: List[String]) {
    import Trie.getNeighbourWithGivenLabel

    private val nodeCounter: Iterator[Int] = Iterator.from(0)
    private val root: Int = nodeCounter.next()
    private val adjacencyList: Map[Node, List[EdgeEndpoint]] = addWordsToTrie(this.words)

    def stringRepresentation: List[String] = {
      (for {
        (node, neighbours) <- adjacencyList
        EdgeEndpoint(neighbour, label) <- neighbours
      } yield s"$node->$neighbour:$label").toList
    }

    private def addWordsToTrie(words: List[String]): Map[Node, List[EdgeEndpoint]] = {
      @tailrec
      def loop(ws: List[String], trie: Map[Node, List[EdgeEndpoint]]): Map[Node, List[EdgeEndpoint]] = ws match {
        case Nil => trie
        case w :: wss =>
          val updatedTrie: Map[Node, List[EdgeEndpoint]] = addWord(trie, w)
          loop(wss, updatedTrie)
      }

      loop(words, Map())
    }

    private def addWord(trie: Map[Node, List[EdgeEndpoint]], w: String): Map[Node, List[EdgeEndpoint]] = {
      @tailrec
      def loop(ls: List[Char],
               acc: Map[Node, List[EdgeEndpoint]],
               currentNode: Node): Map[Node, List[EdgeEndpoint]] = ls match {
        case Nil => acc
        case letter :: lss => getNeighbourWithGivenLabel(acc, currentNode, letter) match {
          case None =>
            val nextNode: Node = nodeCounter.next()
            val updatedTrie: Map[Node, List[EdgeEndpoint]] =
              acc.updated(currentNode, EdgeEndpoint(nextNode, letter) :: acc.getOrElse(currentNode, Nil))
            loop(lss, updatedTrie, nextNode)
          case Some(node) => loop(lss, acc, node)
          }
      }

      loop(w.toList, trie, root)
    }
  }

  object Trie {
    private def getNeighbourWithGivenLabel(trie: Map[Node, List[EdgeEndpoint]],
                                           currentNode: Node,
                                           letter: Char): Option[Node] = {
      trie
        .get(currentNode)
        .flatMap(_.find{ case EdgeEndpoint(_, label) => label == letter })
        .map(_.tip)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrPatterns: Int = reader.next().toInt
    val t0 = System.nanoTime()
    val patterns: List[String] = reader.take(nrPatterns).toList
    val trie = new Trie(patterns)
    trie.stringRepresentation.foreach(println)
  }
}

File: OCaml/string_algorithms/suffix_tries/trie_construction.ml
module IntMap = Map.Make(
    struct
        type t = int
        let compare = Pervasives.compare
    end)

module Trie : sig
        type t
        val create : string list -> t
        val to_string_list : t -> string list
    end = struct
        open Batteries

        type node = int
        type edge_endpoint = { tip : node; label : char }
        type t = edge_endpoint list IntMap.t

        let root = 0

        let get_neighbour_with_given_label trie current_node letter =
            Option.map (function { tip } -> tip ) @@
                Option.bind (IntMap.find_opt current_node trie) (List.find_opt (function { label } -> label = letter ))

        let add_word trie w node_counter =
            let rec loop acc current_node = function
                | [] -> acc
                | letter :: lss -> match (get_neighbour_with_given_label acc current_node letter) with
                    | None ->
                        let next_node = !node_counter in
                        node_counter := !node_counter + 1;
                        let children = match IntMap.find_opt current_node acc with
                            | None -> []
                            | Some values -> values in
                        let updated_trie = IntMap.add current_node ({ tip = next_node; label = letter } :: children) acc
                        in loop updated_trie next_node lss
                    | Some node -> loop acc node lss
            in loop trie root (String.to_list w)

        let create words =
            let node_counter = ref 1 in
            let rec loop trie = function
                | [] -> trie
                | w :: wss ->
                    let updated_trie = add_word trie w node_counter
                    in loop updated_trie wss
            in loop IntMap.empty words

        let to_string_list trie =
            let to_string node neighbour label =
                (string_of_int node) ^ "->" ^ (string_of_int neighbour) ^ ":" ^ (Char.escaped label) in
            List.map
                (function (node, neighbours) ->
                    List.map (function { tip = neighbour; label} -> to_string node neighbour label) neighbours)
                (IntMap.bindings trie) |> List.flatten
    end


let read_patterns n =
    let rec loop acc k =
        if k = n then List.rev acc
        else loop (read_line() :: acc) (k + 1)
    in loop [] 0


let () =
    let number_of_patterns = read_int() in
    let patterns = read_patterns number_of_patterns in
    let trie = Trie.create patterns
    in List.iter print_endline (Trie.to_string_list trie)
